[홈 지도 화면 및 기록 저장 화면] 지도상에 장소와 마커 클릭 기능 구현 및 기록 저장 확인 버튼 클릭시 로컬과 원격 DB에 연동

jungkeunking opened the pull request:
** 2+ 브랜치 작업량을 한 브랜치에서 계속 작업해버리는 큰 실수를 해서 리뷰하기 힘드실 텐데 다시 한번 양해 부탁드립니다

홈 지도 화면
- [x] 장소 클릭시 다이얼로그 표시
- [x] 네트워크 연결 콜백 레포지토리 추가 및 홈 뷰모델에서 연결 상태 관리

홈 지도 화면 다이얼로그
- [x] "네, 가봤어요" 클릭시, 기록 저장 화면으로 이동
- [x] "네, 가볼래요" 클릭시, 장소를 plan = true로 로컬, 원격 DB에 추가
- [x] "아니오" 클릭시, 뒤로가기 

기록 저장 화면
- [x] 확인 버튼 클릭시, plan인 장소의 경우 (닫힌 보물상자 마커 클릭을 통해 현재 화면에 유입), 장소를 plan = false로 로컬, 원격 DB에 업데이트
- [x] 확인 버튼 클릭시, plan이 아닌 장소의 경우 (마커가 없는 지도 심볼 클릭을 통해 현재 화면에 유입), 장소를 로컬, 원격 DB에 추가
- [x] 확인 버튼 클릭시, 기록을 로컬, 원격 DB에 추가
- [x] 확인 버튼 클릭시, 유저를 place, plan, log에 대해 원격 DB에 업데이트

데이터 클래스 관련
- [x] Place, Log에 대해 UI 레이어용 Model 객체 추가

피드백 반영
- [x] 닫힌 보물 상자 마커 클릭 후 기록 저장 화면으로 이동한 다음, 취소 버튼 클릭해 이전 홈 화면으로 이동시에 uid 값이 없어서 마커 표시가 안 되는 오류 해결 -> 홈 지도 화면, 지도 다이얼로그, 기록 저장 화면에서 uid 관리 방식 수정 (navArgs로 전달 -> Firebase.auth로 접근)
- [x] 기록 상세 화면 편집 버튼 클릭시 강종 오류 해결 -> 내비게이션 출발지 변경
- [x] fix: 홈 지도 화면에서 열린 보물상자 마커 클릭해 기록 상세화면으로 이동후 편집 버튼 클릭해 기록 저장화면으로 이동후 취소 버튼 클릭해 다시 홈 지도 화면으로 이동시, 보물상자 마커들이 표시되지 않는 오류 해결  -> 홈 뷰모델에서 로컬 장소 데이터 수집시, merge -> combine 함수 변경 & stateIn 함수 추가해 StateFlow 타입 유지

jungkeunking committed:
build: safeargs, parcelize 의존성 추가

jungkeunking committed:
feat: 홈 지도 화면 장소 클릭시 다이얼로그 표시 구현

ToDo:
- 다이얼로그 수정
- 마커 클릭
- 테스트용 스플래시 -> 홈 화면 전환 되돌리기

jungkeunking committed:
feat: 네트워크 연결 콜백을 처리하는 레포지토리 생성 후 홈 뷰모델에 추가 및 MapUiState에 연결 상태 프로퍼티 추가

jungkeunking committed:
feat: 장소를 원격 DB에 추가하는 기능 구현

jungkeunking committed:
feat: 홈 지도 화면 다이얼로그창에서 기록 저장 화면으로 이동 및 데이터 전달 구현

jungkeunking committed:
feat: 기록 저장 화면에서 장소를 로컬, 원격 DB에 추가하고 기록을 로컬 DB에 추가하는 기능 구현

jungkeunking committed:
refactor: 홈 지도 화면 함수명 변경 및 스플래시 화면 테스트용 변경사항 복구

jungkeunking committed:
fix: 홈 지도 화면에서 마커 클릭시 바텀시트 표시하는 함수 오류 수정

jungkeunking committed:
fix: PlaceDao에서 update 함수 suspend로 변경 및 레포지토리에 적용

메인 쓰레드를 블록킹하는 오류 수정

jungkeunking committed:
feat: 홈 지도 화면에서 place 및 plan 보물상자 마커 클릭 기능 구현

jungkeunking committed:
feat: 기록 저장 화면에서 확인 버튼 클릭해 장소 저장시 원격 id 업데이트

jungkeunking committed:
refactor: 지도 다이얼로그 스트링 리소스 추가 및 함수 리팩토링

jungkeunking committed:
fix: 홈 지도 화면에 가볼 곳 (plan) 표시 오류 수정 및 addPlan -> addPlace 함수명 변경

jungkeunking committed:
feat: 스플래시 화면 및 로그인 화면에서 로그인 후 화면 이동시 uid 전달

jungkeunking committed:
feat: 원격 데이터 업데이트 함수 추가

jungkeunking committed:
feat: 홈 화면에서 plan 마커 클릭 후 화면 이동시 uid 및 remotePlaceId 전달

jungkeunking committed:
feat: 홈 지도 화면 다이얼로그에서 plan 추가시 로컬 및 원격 DB에 저장

jungkeunking committed:
feat: 기록 저장 화면에서 기존 plan인지 처음 place인지 구분해 장소, 기록 데이터 로컬 및 원격 DB에 저장

jungkeunking committed:
feat & refactor: uid가 null일 경우 마커 표시하는 함수 리턴 및 코드 리팩토링

jungkeunking committed:
refactor: 불필요한 코드 삭제 및 수정

jungkeunking committed:
refactor: 코드 리팩토링

jungkeunking committed:
refactor: 현재 시간 반환하는 코드를 유틸 함수로 분리

jungkeunking committed:
refactor: UI 모델에 해당하는 클래스들 패키지 분리

jungkeunking committed:
feat: Log, Place에 대해 UI 모델 및 Entity, DTO로 변환 함수 추가

jungkeunking committed:
feat: Place에 대해 Entity -> DTO 변환 함수 추가 및 프로퍼티 순서 수정

jungkeunking committed:
feat: insert LogDTO 함수를 스트링 id 리턴하도록 수정 및 Entity -> DTO 변환 함수 추가

jungkeunking committed:
fix: UserDTO에 대해 logs, places, plans 프로퍼티를 리스트 -> 맵 타입으로 수정

jungkeunking committed:
refactor: UI 모델 패키지 분리 및 UserDTO 프로퍼티들 맵 타입으로 변경 사항들 적용

jungkeunking committed:
feat & refactor: 유저 데이터를 원격에서 업데이트하는 함수 추가 및 사진 업로드 로직을 뷰모델로 이전

jungkeunking committed:
feat: 기록 저장 화면에서 확인 버튼 클릭시, 장소, 기록, 유저 데이터를 로컬 및 원격 DB에 업데이트

jungkeunking committed:
fix: 로그인 뷰모델 테스트 샘플 제거

jungkeunking committed:
fix: 홈 지도 화면, 지도 다이얼로그, 기록 저장 화면에서 uid 관리 방식 수정 (navArgs로 전달 -> Firebase.auth로 접근)

닫힌 보물 상자 마커 클릭 후 기록 저장 화면으로 이동한 다음, 취소 버튼 클릭해 이전 홈 화면으로 이동시에 uid 값이 없어서 마커 표시가 안 되는 오류 해결

jungkeunking committed:
feat: 홈 지도 화면에서 마커 표시시 네트워크 연결 확인 로직 추가 및 로그 함수 제거

donghyun81 commented on app/src/main/java/com/treasurehunt/ui/home/HomeFragment.kt:
```diff
@@ -147,27 +142,42 @@ class HomeFragment : Fragment(), OnMapReadyCallback {
+                        marker.map = map
+                        marker.setPlaceClick(remotePlaceId)
+                    }
+                    uiState.planMarkers.forEach { marker ->
```
예전에 forEach보다 map을 권장한다고 하셨던거 같은데 forEach로 사용하신 이유가 있을까요??

jungkeunking replied:
음 제가 어느 맥락에서 그 말씀을 드렸는지 기억이 안나는데, 반환값을 활용하지 않는 경우였었을까요?
저도 forEach를 선호하지는 않아서 음 의도를 더 드러내는 고차함수 있을지 한번더 고민해보겠습니다

donghyun81 replied:
제가 저번에 forEach를 사용했을때 반환값이 없어서 forEach로 사용했다고 했을때 map을 추천해주셨는데 피드백 내용이 강사님이 의도 관련해서 피드백 주신적이 있다 였던거 같아서 질문 드렸습니다!

jungkeunking replied:
음 그 경우는 데이터를 맵핑하는 거였을 거라고 추측되구요
여기서는 마커들을 다른 객체로 맵핑하는게 아니라 순회하는 거라 해당이 안될 거 같아요

jungkeunking committed:
fix: 기록 상세 화면 편집 버튼 클릭시 강종 오류 해결

donghyun81 commented on app/src/main/java/com/treasurehunt/ui/home/HomeViewModel.kt:
```diff
@@ -75,14 +74,14 @@ class HomeViewModel(
-                placeRepo.getAllPlans().collect { plans ->
-                    _uiState.update {
-                        it.copy(planMarkers = plans.mapToMarkers())
+                merge(placeRepo.getAllPlaces(), placeRepo.getAllPlans()).collect { placesAndPlans ->
```
코드를 봤을때  merge를 combine으로 변경해서 각 flow의 값을 원하는 방식으로 합치는게 구현 내용이 더 좋아질꺼 같은데 확인 해주시겠어요? 
https://github.com/android/architecture-samples/blob/main/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt 의 66번줄 코드 확인해보시면 됩니다!

jungkeunking replied:
동현님이 선견지명이 있으셨네요
오류 해결 과정에서 combine이 열쇠였습니다 
그런데 제가 아직 merge, combine 차이에 대해 학습 중인데, 동현님은 어떤 이유로 combine이 구현 내용이 더 좋아질 거라고 판단하셨나요?

donghyun81 replied:
제가 이해한거는 merge는 여러 flow의 값을 처리 순서데로 가져와서 병합하고
combine은 각 flow의 처리 값을 사용자가 정의한 방식으로 병합 할 수 있어서
코드중에 merge를 통해 가져온 처리를 여러 함수를 통해 필요한 값을 선별하시는 과정 때문에 추천드렸습니다!

jungkeunking replied:
음 제가 찾아본 내용도 그 정도인데, 사실 저는 두 데이터를 결합할 때 특별한 변환을 해준다기보다 그냥 더하는 거라서 그 이유가 해당되는지는 잘 모르겠어요
좀더 공부해봐야할 거 같습니다!

donghyun81 commented on app/src/main/java/com/treasurehunt/ui/savelog/SaveLogFragment.kt:
```diff
@@ -133,111 +137,116 @@ class SaveLogFragment : Fragment(), OnMapReadyCallback {
+        }
+
+        return remotePlaceId
+    }
```
if문 안에 있는 변경점들 함수로 분리하면 더 보기 좋을거 같아요

jungkeunking replied:
훌륭한 조언입니다!!
반영했습니당

donghyun81 reviewed (commented):
고생하셨습니다!

jungkeunking committed:
refactor: 기록상세화면으로 이동시 젯팩 내비게이션 적용

jsulim commented on app/src/main/java/com/treasurehunt/ui/savelog/SaveLogFragment.kt:
마커 클릭해서 열린 보물상자와 닫힌 보물상자를 여러 개 만들고 닫힌 보물상자를  클릭하여 기록 저장 화면으로 넘어갔다가 게시물을 작성하지 않고 다시 뒤로가기를 누르면 닫힌 보물상자 목록들이 모두 사라지는데 이건 데이터 전달이 안된 상태여서 그런걸까요? 아니면 오류일까요?

jungkeunking replied:
중요한 오류인데 잘 집어내주셨네요~~
가볼 장소 (plan), 가본 장소 (visit) 각 플로우 데이터를 결합해 수집하는 과정에서 문제가 있었던 걸로 판명되고, 결합하는 함수를 merge -> combine으로 변경해 오류 해결했습니다
정확한 이유는 좀 더 조사 중입니다!

jsulim commented:
구현을 정말 많이 하셨네요 고생하셨습니다!

jungkeunking committed:
fix: 홈 뷰모델에서 로컬 장소 데이터 수집시, merge -> combine 함수 변경

홈 지도 화면에서 열린 보물상자 마커 클릭해 기록 상세화면으로 이동후 편집 버튼 클릭해 기록 저장화면으로 이동후 취소 버튼 클릭해 다시 홈 지도 화면으로 이동시, 보물상자 마커들이 표시되지 않는 오류 해결

jungkeunking committed:
feat: 홈 뷰모델 플로우 데이터 결합시 stateIn 추가해 StateFlow 타입 유지 및 기록 저장 화면, 뷰모델 import 정리

jungkeunking committed:
refactor: 홈 뷰모델에서 combine 함수를 확장 -> 일반함수로 변경

jungkeunking committed:
refactor: 기록저장 화면에서 insertOrUpdatePlace 함수 리팩토링 및 닫힌 보물상자 마커에 remoteId 값 부재시 오류 스낵바 표시

jungkeunking committed:
fix: LogModel.asLogDTO 함수 오류 수정

jungkeunking committed:
fix: 저장화면 뷰모델 uploadTask.await 함수 임시 주석 처리

retrofit2.HttpException: HTTP 400 Bad Request 오류 발생
스낵바 구현 로직이 변경될 있으므로 디버깅 보류

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/savelog/SaveLogFragment.kt:
```diff
@@ -206,38 +216,39 @@ class SaveLogFragment : Fragment(), OnMapReadyCallback {
-            lng,
-            false,
-            caption
+    private suspend fun getRemoteAndLocalPlaceIdFromPlanMarker(mapSymbol: MapSymbol): Pair<String, Long>? {
+        val remotePlaceId: String = mapSymbol.remoteId ?: return null
+        val localPlaceId = viewModel.getRemotePlaceById(remotePlaceId).id
+
+        return (remotePlaceId to localPlaceId)
+    }
```
뭔가 함수 이름이 너무 긴 것 같습니다. and랑 from 이런게 함수 이름에 되도록이면 사용이 덜 되는게 좋아보이는데 그냥 getPlaceId는 어떤가요?

jungkeunking replied:
음 저도 고민 많이 한 부분인데, 저게 깔끔해 보여서 수정하겠습니다!
이제 보니 리턴타입을 통해 아이디 두개를 반환한다고 유추할 수 있겠네요

juhan2103 reviewed (commented):
@jungkeunking 구현 수고하셨습니다! 예외처리 부분이라던지 코드를 줄이려고 노력하신 내용이 보여 엄청나다고 생각했습니다! 코드리뷰 하나 달았는데 확인부탁드립니다!

jungkeunking committed:
refactor: 함수명 변경
